sdk/core/Azure.Core/src/RetryOptions.cs

```diff
@@ -37,29 +50,85 @@ internal RetryOptions(RetryOptions? retryOptions)
         /// <summary>
         /// The maximum number of retry attempts before giving up.
         /// </summary>
-        public int MaxRetries { get; set; } = 3;
+        public int MaxRetries
+        {
+            get => _maxRetries;
+            set
+            {
+                AssertNotFrozen();
+
+                _maxRetries = value;
+            }
+        }
 
         /// <summary>
         /// The delay between retry attempts for a fixed approach or the delay
         /// on which to base calculations for a backoff-based approach.
         /// If the service provides a Retry-After response header, the next retry will be delayed by the duration specified by the header value.
         /// </summary>
-        public TimeSpan Delay { get; set; } = TimeSpan.FromSeconds(0.8);
+        public TimeSpan Delay
+        {
+            get => _delay;
+            set
+            {
+                AssertNotFrozen();
+
+                _delay = value;
+            }
+        }
 
         /// <summary>
         /// The maximum permissible delay between retry attempts when the service does not provide a Retry-After response header.
         /// If the service provides a Retry-After response header, the next retry will be delayed by the duration specified by the header value.
         /// </summary>
-        public TimeSpan MaxDelay { get; set; } = TimeSpan.FromMinutes(1);
+        public TimeSpan MaxDelay
+        {
+            get => _maxDelay;
+            set
+            {
+                AssertNotFrozen();
+
+                _maxDelay = value;
+            }
+        }
 
         /// <summary>
         /// The approach to use for calculating retry delays.
         /// </summary>
-        public RetryMode Mode { get; set; } = RetryMode.Exponential;
+        public RetryMode Mode
+        {
+            get => _retryMode;
+            set
+            {
+                AssertNotFrozen();
+
+                _retryMode = value;
+            }
+        }
 
         /// <summary>
         /// The timeout applied to an individual network operations.
         /// </summary>
-        public TimeSpan NetworkTimeout { get; set; } = TimeSpan.FromSeconds(100);
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public TimeSpan NetworkTimeout
+        {
+            get => _networkTimeout;
+            set
+            {
+                AssertNotFrozen();
+
+                _networkTimeout = value;
+            }
+        }
+
+        internal void Freeze() => _frozen = true;
+
+        private void AssertNotFrozen()
+        {
+            if (_frozen)
+            {
+                throw new InvalidOperationException("Cannot change a ClientOptions instance after it has been used to create a client.");
```

Comment:
Is this exception targeted at end users or client developers? How would end users reason about how to avoid hitting this?